File: day04.py
with open('day04_input.txt') as f:
    input = f.read().splitlines()
# print(input)

range_pairs = []
for range_pair in input:
    range1, range2 = range_pair.split(",")
    start1, end1 = range1.split("-")
    start2, end2 = range2.split("-")
    range_pairs.append({
        'start1': int(start1),
        'end1': int(end1),
        'start2': int(start2),
        'end2': int(end2)})
# print(range_pairs)


def find_smaller(range_pair):
    range1 = range_pair["end1"] - range_pair["start1"]
    range2 = range_pair["end2"] - range_pair["start2"]
    if range1 >= range2:
        return 2
    else:
        return 1


def check_range(range_pair):
    smaller = find_smaller(range_pair)
    if smaller == 1:
        return start1 >= start2 and end1 <= end2
    else:
        return start2 >= start1 and end2 <= end1


answer = 0
for range_pair in range_pairs:
    if check_range(range_pair):
        print(f'{range_pair} is in range')
        answer += 1
    else:
        print(f'{range_pair} is not in range')
print(answer)

File: day04.js
const input = require("fs")
  .readFileSync("day04_input.txt")
  .toString()
  .split(/\r?\n/)

let rangePairs = []

input.forEach((pair) => {
  let [range1, range2] = pair.split(",")
  let [range1Start, range1End] = range1.split("-")
  let [range2Start, range2End] = range2.split("-")
  rangePairs.push({
    start1: parseInt(range1Start),
    end1: parseInt(range1End),
    start2: parseInt(range2Start),
    end2: parseInt(range2End),
  })
})

let answer = 0
rangePairs.forEach((pair) => {
  if (rangeCheck(pair)) {
  }
})

console.log({ answer })
function rangeCheck(rangePair) {
  let { start1, end1, start2, end2 } = rangePair
  let smaller = findSmaller(rangePair)
  if (smaller === 1) {
    return start1 >= start2 && end1 <= end2
  } else {
    return start2 >= start1 && end2 <= end1
  }
}

function findSmaller(rangePair) {
  let { start1, end1, start2, end2 } = rangePair
  let range1 = parseInt(end1 - start1)
  let range2 = parseInt(end2 - start2)
  if (range1 >= range2) {
    return 2
  } else {
    return 1
  }
}
